lib/routes/clients-instances.js

```diff
@@ -0,0 +1,341 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+const {URL} = require('url')
+const AppError = require('../error');
+const error = require('../error')
+const isA = require('joi')
+const ScopeSet = require('fxa-shared').oauth.scopes
+
+// Assign a default TTL for well-known commands if a request didn't specify it.
+const DEFAULT_COMMAND_TTL = new Map([
+  ['https://identity.mozilla.com/cmd/open-uri', 30 * 24 * 3600], // 30 days
+])
+
+module.exports = (log, db, config, customs, clients, pushbox, push) => {
+  const routes = [
+    {
+      method: 'GET',
+      path: '/clients_instances',
+      options: {
+        auth: {
+          strategies: [
+            'sessionToken',
+            'oauthToken'
+          ]
+        },
+        response: {
+          schema: isA.array().items(isA.object({
+            id: isA.string().required(),
+            clientId: isA.string().required(),
+            name: isA.string().required().allow(null),
+            pushEndpoint: isA.string().required().allow(null),
+            pushPublicKey: isA.string().required().allow(null),
+            pushAuthKey: isA.string().required().allow(null),
+            availableCommands: isA.object().required(),
+          }))
+        }
+      },
+      async handler(request) {
+        log.begin('ClientsInstances', request)
+        if (request.auth.strategy === 'oauthToken' &&
+            ! ScopeSet.fromArray(request.auth.credentials.scope).contains('clients:read')) {
```

Comment:
We might already be using the "clients" scope namespace for adding/removing client_ids in the oauth-server.